mautic/mautic
# Reports improvements

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/4992 https://github.com/mautic/mautic/issues/4714 https://github.com/mautic/mautic/issues/4704
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
When I prepared reports for clients I noticed some issues and improvements opportunity.

## Bugfixes

**Fix for When adding "unsubscribed" to an emails report the report errors out** 
reported here https://github.com/mautic/mautic/issues/4992

How to test: 

1. Add Emails report with unsubscribe column . 
2. No error = works!

**Added support for Channel ID for unsubscribe through preferences center** 

How to test:

1. Enable Preferences center
2. Send email with unsubscribe link and click to unsubscribe link
3. Unsubscribe email channel, go to DB to lead_donotcontact and check unsubscribe item with null channel id
4. Apply PR and try re-subscribe and unsubscribe again via preferences center. 
5:  go to DB to lead_donotcontact and check  unsubscribe item with regular channel id

For code reviewers,  the changes are here: https://github.com/mautic/mautic/pull/4995/files#diff-943c28947c87cf9426037ee9fda3335aR63

## New table columns

Now we have divided unsubscribed and bounced stats.

Emails report: Bounced, Bounced ratio, Read delay
Emails sent report: Clicked, Bounced, Unsubscribed 

How to test: 

1. Test columns with your data.

![image](https://user-images.githubusercontent.com/462477/30859275-4209e9b2-a2c3-11e7-8ba0-b412213517e0.png)

## New graphs

Emails report: New pie graph:  **Read / Ignored / Unsubscribed / Bounced**
![image](https://user-images.githubusercontent.com/462477/30859462-1d38f140-a2c4-11e7-80bb-efb2d01feb37.png)

Emails sent report: New table graphs:  **Most Emails unsubscribed and Most Emails bounced i**
![image](https://user-images.githubusercontent.com/462477/30859515-5f2ee528-a2c4-11e7-9071-0f56dcbcf795.png)

How to test:

1. Test graphs with your data.

Added new graphs in emails and emails sents reports. See if works with your data

## Dynamic filters improvements
There are several changes.

1.  Bool filter type improvments
![image](https://user-images.githubusercontent.com/462477/30889922-e63422bc-a329-11e7-8fd4-f275a1daa5b4.png)

How to test:

1. Create Emails sent report
2. Add Read, Clicked, Unsubscribed, Bounced columns and filters
3. Enable dynamic filters  everywhere
4. See new Clear Filter option for bool types 
5.  Select random default values (Yes, No, Clear Filter)
6. Go to report 
7. Works If bool dynamic filters are predefined
8. Try change bool dynamic filters. Check If table data are OK.
9. Change order on any columns
10. Go to export HTML
11. Works If you see the same results like on report page

## New settings

Update bring also new "settings" column to mysql table and migration file.
Added 3 new settings to reports

Two settings are in Dynamic filters settings below Group by on Data tab

![image](https://user-images.githubusercontent.com/462477/30890108-17f3ce5a-a32b-11e7-8547-e5a894d9879c.png)

How to test:

1. Create  Emails report, add dynamic filters and aenable both options.
2. Enable Show dynamic filters opened by default - you should see opened dynamic filters area on report page
3. Enable Hide date ranges in dynamic filter area -  you shouldn't find  date range in dynamic filters on report page  (this option is helpful for Emails report - there's no apply date range)

Another one is on Graphs tab

![image](https://user-images.githubusercontent.com/462477/30890175-881dffc0-a32b-11e7-83d2-20ad2429dd4c.png)

How to test::

1. Create  Emails sent report, add columns in data and also add graphs
2. Enable Show dynamic filters opened by default  in Graphs tab
3. Go to report page and see If graphs show above table
4. Disable option
5. Go to report page and see If graphs show below table
6. Try also test with HTML export 

## Change Reports owner
Few times I created reports for another user with very limited access. Now We can switch owners and this user can see just his  reports 

1. Add user with view own reports access
2. Create report and change owner to created user
3: Login as created user and see if you can view report but not others reports
